### PR TITLE
Optimizer State Sharing & Async State Loading

### PR DESCRIPTION
**Pull request by KMFODA**
This PR does the following:

- [x] Converts the model loading function to an async background thread which should reduce timeouts within the Train synapse call
- [x] Normalizes gradient and steps scores
- [x] Uploads a global version of the master validator's optimizer state dictionary to allow any out of sync miners to re-init their optimisers with the correct states

### CODE REVIEW

File: distributed_training/validator/reward.py [hunk 1]
```diff
@@ -370,6 +370,9 @@ async def get_rewards(
                 for uid, gradient_score in zip(uids, gradient_scores)
             }
         )
+        gradient_scores = torch.nn.functional.normalize(
+            torch.tensor([gradient_scores.tolist()]).to(self.device)
+        )
```
[the remark this answers is not in the document]
Comment (KMFODA) [reply distributed_training/validator/reward.py:375]: It isn't no. Let me know what part of the code makes you think this.

File: distributed_training/validator/reward.py [hunk 2]
```diff
@@ -416,11 +422,15 @@ async def get_rewards(
             )
 
             # Final balanced score calculation with all_reduce
-            scores = blacklist_scores * (
-                (0.5 * gradient_scores * steps_scores) + (0.5 * all_reduce_scores)
+            scores = (
+                blacklist_scores
+                * ((0.5 * gradient_scores * steps_scores) + (0.5 * all_reduce_scores))[
```
[the remark this answers is not in the document]
Comment (KMFODA) [reply distributed_training/validator/reward.py:427]: I think that might be too harsh given the nature of hivemind's all_reduce is that it doesn't work perfectly each step in allowing peers to join a group and sometimes even if they join the connection drops out.

[the remark this answers is not in the document]
Comment (KMFODA) [reply distributed_training/validator/reward.py:427]: The average of a multiplicative approach though might end up favouring honest miners over malicious ones so it could end up being a better reward mechanism. We'd need to test this to be sure.

File: distributed_training/validator/reward.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (KMFODA) [reply distributed_training/validator/reward.py:375]: Ah this is normalising all the lists of gradients we've compared for a single miner. Whereas here we're normalising all those miners scores against each other.

[the remark this answers is not in the document]
Comment (KMFODA) [reply distributed_training/validator/reward.py:375]: It might be redundant to have both though, you're right.
